File: frontend/types/events.ts
/**
 * Types pour les événements SSE du backend Agent4BA
 */

export interface TestCaseStep {
  step: string;
  expected_result: string;
}

export interface TestCase {
  title: string;
  scenario: string;
  steps: TestCaseStep[];
}

export interface WorkItem {
  id: string;
  project_id: string;
  type: "feature" | "story" | "task" | "test_case";
  title: string;
  description: string;
  parent_id: string | null;
  acceptance_criteria?: string[];
  test_cases?: TestCase[];
  validation_status: "ia_generated" | "human_validated" | "ia_modified";
  attributes: {
    priority?: string;
    status?: string;
    points?: number;
    [key: string]: unknown;
  };
}

export interface ContextItem {
  type: "document" | "work_item";
  id: string;
  name: string;
}

export interface ModifiedItem {
  before: WorkItem;
  after: WorkItem;
}

export interface ImpactPlan {
  new_items: WorkItem[];
  modified_items: ModifiedItem[];
  deleted_items: string[];
}

// Base type for all SSE events
export interface BaseEvent {
  type: string;
}

export interface ThreadIdEvent extends BaseEvent {
  type: "thread_id";
  thread_id: string;
}

export interface UserRequestEvent extends BaseEvent {
  type: "user_request";
  query: string;
}

export interface NodeStartEvent extends BaseEvent {
  type: "node_start";
  node_name: string;
}

export interface NodeEndEvent extends BaseEvent {
  type: "node_end";
  node_name: string;
  output?: Record<string, unknown>;
}

export interface LLMStartEvent extends BaseEvent {
  type: "llm_start";
  model?: string;
}

export interface LLMTokenEvent extends BaseEvent {
  type: "llm_token";
  token: string;
}

export interface LLMEndEvent extends BaseEvent {
  type: "llm_end";
  content?: string;
}

export interface ImpactPlanReadyEvent extends BaseEvent {
  type: "impact_plan_ready";
  impact_plan: ImpactPlan;
  thread_id: string;
  status: string;
}

export interface WorkflowCompleteEvent extends BaseEvent {
  type: "workflow_complete";
  result: string;
  status: string;
}

export interface ErrorEvent extends BaseEvent {
  type: "error";
  error: string;
  details?: string;
}

export interface AgentStartEvent extends BaseEvent {
  type: "agent_start";
  thought: string;
  agent_name: string;
}

export interface AgentPlanEvent extends BaseEvent {
  type: "agent_plan";
  steps: string[];
  agent_name: string;
}

export interface ToolUsedEvent extends BaseEvent {
  type: "tool_used";
  tool_run_id: string;
  tool_name: string;
  tool_icon: string;
  description: string;
  status: "running" | "completed" | "error";
  details?: Record<string, unknown>;
}

// Structure pour organiser les événements par session
export interface TimelineSession {
  id: string;
  user_query: string;
  timestamp: Date;
  tool_runs: Map<string, ToolRunState>;
  agent_events: TimelineEvent[];
  is_expanded: boolean;
}

// État d'une exécution d'outil avec ses mises à jour
export interface ToolRunState {
  tool_run_id: string;
  tool_name: string;
  tool_icon: string;
  description: string;
  status: "running" | "completed" | "error";
  details: Record<string, string | number | boolean | null | undefined>;
  started_at: Date;
  completed_at?: Date;
}

// Union type of all possible events
export type SSEEvent =
  | ThreadIdEvent
  | UserRequestEvent
  | NodeStartEvent
  | NodeEndEvent
  | LLMStartEvent
  | LLMTokenEvent
  | LLMEndEvent
  | ImpactPlanReadyEvent
  | WorkflowCompleteEvent
  | ErrorEvent
  | AgentStartEvent
  | AgentPlanEvent
  | ToolUsedEvent;

// For timeline display
export interface TimelineEvent {
  id: string;
  timestamp: Date;
  event: SSEEvent;
}

// Multi-turn conversation types
export interface ClarificationNeededResponse {
  status: "clarification_needed";
  conversation_id: string;
  question: string;
}

export interface ApprovalNeededResponse {
  status: "awaiting_approval";
  thread_id: string;
  impact_plan: ImpactPlan;
}

export interface ExecuteSuccessResponse {
  result: string;
  project_id: string;
  status: string;
}

export interface RespondSuccessResponse {
  result: string;
  project_id: string;
  status: string;
  thread_id: string | null;
  impact_plan: ImpactPlan | null;
}

File: agent4ba/api/events.py
"""
Schémas d'événements pour le streaming SSE.

Ce module définit les structures de données pour les événements
envoyés au client via Server-Sent Events (SSE).
"""

from typing import Any

from pydantic import BaseModel, Field


class StreamEvent(BaseModel):
    """Base class pour tous les événements SSE."""

    type: str = Field(..., description="Type de l'événement")


class ThreadIdEvent(StreamEvent):
    """Événement envoyé au début du stream avec l'identifiant de thread."""

    type: str = Field(default="thread_id", description="Type d'événement")
    thread_id: str = Field(..., description="Identifiant unique du thread de conversation")


class UserRequestEvent(StreamEvent):
    """Événement contenant la requête initiale de l'utilisateur."""

    type: str = Field(default="user_request", description="Type d'événement")
    query: str = Field(..., description="Requête de l'utilisateur")


class NodeStartEvent(StreamEvent):
    """Événement indiquant le début d'exécution d'un nœud."""

    type: str = Field(default="node_start", description="Type d'événement")
    node_name: str = Field(..., description="Nom du nœud qui démarre")


class NodeEndEvent(StreamEvent):
    """Événement indiquant la fin d'exécution d'un nœud."""

    type: str = Field(default="node_end", description="Type d'événement")
    node_name: str = Field(..., description="Nom du nœud qui se termine")
    output: dict[str, Any] | None = Field(
        default=None,
        description="Données de sortie du nœud",
    )


class LLMStartEvent(StreamEvent):
    """Événement indiquant le début d'un appel LLM."""

    type: str = Field(default="llm_start", description="Type d'événement")
    model: str | None = Field(None, description="Nom du modèle LLM utilisé")


class LLMTokenEvent(StreamEvent):
    """Événement contenant un token streamé du LLM."""

    type: str = Field(default="llm_token", description="Type d'événement")
    token: str = Field(..., description="Token généré par le LLM")


class LLMEndEvent(StreamEvent):
    """Événement indiquant la fin d'un appel LLM."""

    type: str = Field(default="llm_end", description="Type d'événement")
    content: str | None = Field(None, description="Contenu complet généré par le LLM")


class ImpactPlanReadyEvent(StreamEvent):
    """Événement envoyé quand un ImpactPlan est prêt pour validation."""

    type: str = Field(default="impact_plan_ready", description="Type d'événement")
    impact_plan: dict[str, Any] = Field(..., description="Plan d'impact généré")
    thread_id: str = Field(..., description="Identifiant du thread pour validation")
    status: str = Field(
        default="awaiting_approval",
        description="Statut du workflow",
    )


class SchemaChangeReadyEvent(StreamEvent):
    """Événement envoyé quand un changement de schéma est prêt pour validation."""

    type: str = Field(default="schema_change_proposed", description="Type d'événement")
    proposed_schema: dict[str, Any] = Field(..., description="Nouveau schéma proposé")
    thread_id: str = Field(..., description="Identifiant du thread pour validation")
    status: str = Field(
        default="awaiting_schema_approval",
        description="Statut du workflow",
    )


class WorkflowCompleteEvent(StreamEvent):
    """Événement indiquant la fin du workflow."""

    type: str = Field(default="workflow_complete", description="Type d'événement")
    result: str = Field(..., description="Résultat final du workflow")
    status: str = Field(..., description="Statut final")


class ErrorEvent(StreamEvent):
    """Événement indiquant une erreur pendant l'exécution."""

    type: str = Field(default="error", description="Type d'événement")
    error: str = Field(..., description="Message d'erreur")
    details: str | None = Field(None, description="Détails supplémentaires sur l'erreur")


class AgentStartEvent(StreamEvent):
    """Événement émis au début de l'exécution d'un agent avec sa reformulation."""

    type: str = Field(default="agent_start", description="Type d'événement")
    thought: str = Field(..., description="Reformulation de ce que l'agent a compris et va faire")
    agent_name: str = Field(..., description="Nom de l'agent qui démarre")


class AgentPlanEvent(StreamEvent):
    """Événement contenant le plan d'action de l'agent."""

    type: str = Field(default="agent_plan", description="Type d'événement")
    steps: list[str] = Field(..., description="Liste des étapes prévues par l'agent")
    agent_name: str = Field(..., description="Nom de l'agent")


class ToolUsedEvent(StreamEvent):
    """Événement indiquant l'utilisation d'un outil par l'agent."""

    type: str = Field(default="tool_used", description="Type d'événement")
    tool_run_id: str = Field(..., description="Identifiant unique de cette exécution d'outil (UUID)")
    tool_name: str = Field(..., description="Nom de l'outil utilisé")
    tool_icon: str = Field(..., description="Emoji/icône représentant l'outil")
    description: str = Field(..., description="Description de ce que fait l'outil")
    status: str = Field(default="running", description="Statut: running, completed, error")
    details: dict[str, Any] | None = Field(None, description="Détails additionnels sur l'utilisation de l'outil")
